datasetinsights/estimators/faster_rcnn.py

```diff
@@ -22,6 +22,605 @@
+                logger.info(
+                    f"intermediate loss after mini batch {i} in epoch {epoch} "
+                    f"(total training examples: {examples_seen}) is "
+                    f"{intermediate_loss}"
+                )
+                self.writer.add_scalar(
+                    "training/running_loss", intermediate_loss, examples_seen
+                )
+            losses_grad.backward()
+            if (i + 1) % accumulation_steps == 0:
+                optimizer.step()
+                lr_scheduler.step()
+                optimizer.zero_grad()
+        self.writer.add_scalar("training/loss", loss_metric.compute(), epoch)
+        self.writer.add_scalar(
+            "training/running_loss",
+            loss_metric.compute(),
+            loss_metric.num_examples,
+        )
+        self.writer.add_scalar(
+            "training/lr", optimizer.param_groups[0]["lr"], epoch
+        )
+        loss_metric.reset()
+
+    def evaluate(self, **kwargs):
+        config = self.config
+        test_dataset = download_data(config, TEST)
+        label_mappings = test_dataset.label_mappings
+        if self.config.system.dryrun:
+            test_dataset = create_dryrun_dataset(config, test_dataset, TEST)
+
+        is_distributed = config.system.distributed
+        test_sampler = create_sampler(
+            is_distributed=is_distributed, dataset=test_dataset, is_train=False
+        )
+
+        logger.info(f"length of test dataset is {len(test_dataset)}")
+        logger.info("Start evaluating estimator: %s", type(self).__name__)
+
+        test_loader = create_dataloader(
+            config, test_dataset, test_sampler, TEST
+        )
+        self.model.to(self.device)
+        self.evaluate_per_epoch(
+            data_loader=test_loader,
+            epoch=0,
+            label_mappings=label_mappings,
+            is_distributed=self.config.system.distributed,
+            synchronize_metrics=self.sync_metrics,
+        )
+
+    @torch.no_grad()
+    def evaluate_per_epoch(
+        self,
+        *,
+        data_loader,
+        epoch,
+        label_mappings,
+        max_detections_per_img=MAX_BOXES_PER_IMAGE,
+        is_distributed=False,
+        synchronize_metrics=True,
+    ):
+        """
+        Evaluate model performance. Note, torchvision's implementation of faster
+        rcnn requires input and gt data for training mode and returns a
+        dictionary of losses (which we need to record the loss). We also need to
+        get the raw predictions, which is only possible in model.eval() mode,
+        to calculate the evaluation metric.
+        Args:
+            data_loader (DataLoader): pytorch dataloader
+            epoch (int): current epoch, used for logging
+            label_mappings (dict): a dict of {label_id: label_name} mapping
+            max_detections_per_img: max number of targets or predictions allowed
+            per example
```

Comment:
[nit] might be better to indent this line to improve readability.

Reply:
[nit]?

Reply:
Just mean this is nit-picking comment in the context of PR

https://stackoverflow.com/questions/27810522/what-does-nit-mean-in-hacker-speak